crates/cards/src/validate.rs

```diff
@@ -0,0 +1,88 @@
+use std::{fmt, str::FromStr};
+
+use masking::{Strategy, StrongSecret, WithType};
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+#[error("not a valid credit card number")]
+pub struct CCValError;
+
+impl From<core::convert::Infallible> for CCValError {
+    fn from(_: core::convert::Infallible) -> Self {
+        Self
+    }
+}
+
+/// Card number
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct CardNumber(StrongSecret<String>);
+
+impl FromStr for CardNumber {
+    type Err = CCValError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match luhn::valid(s) {
+            true => Ok(Self(StrongSecret::from_str(s)?)),
+            false => Err(CCValError),
+        }
+    }
+}
+
+impl<T> Strategy<T> for CardNumber
+where
+    T: AsRef<str>,
+{
+    fn fmt(val: &T, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let val_str: &str = val.as_ref();
+
+        if val_str.len() < 15 || val_str.len() > 19 {
+            return WithType::fmt(val, f);
+        }
+
+        write!(f, "{}{}", &val_str[..6], "*".repeat(val_str.len() - 6))
+    }
+}
```

Comment:
Hey, you might have had a slight misunderstanding regarding how a masking `Strategy` is supposed to function. I'll briefly explain the relationship between the `Data` and the `Strategy` in `masking::Secret<Data, Strategy>` in case there's some confusion.

The `Data` type parameter is a type that represents the sensitive data, and the `Strategy` type parameter represents the strategy using which the `Data` is to be masked. All the data is stored in `Data`, the `Strategy` is just intended to be a simple (usually a unit struct) type that provides an impl of the `Strategy<T>::fmt` method (which doesn't take any `&self`).

Here, your `CardNumber` struct is doing both, storing a card number `Data` and providing the `Strategy<T>` impl. This means that when you specify `masking::Secret<String, CardNumber>`, the `String` is storing your card number, and `CardNumber` just provides the strategy even tho it's capable of doing both, along with providing validations.

Since you're trying to do validations, my suggestion would be to implement a separate unit struct `CardNumberStrategy` which provides the `Strategy<T>` impl, updating the `CardNumber` struct to `struct CardNumber(StrongSecret<String, CardNumberStrategy>)` and update usages of `masking::Strategy<String, CardNumber>` in the code with just `cards::CardNumber`.
With this, you have a wrapper type (`CardNumber`) around a `StrongSecret` storing your card number which provides validations, and internally, the `StrongSecret` holds the masking strategy.

If you're going down this route, also consider implementing `Deref` on `CardNumber`.